### ISSUE TITLE
Docusaurus Faster

### ISSUE THREAD

**Issue by slorber** [facebook/docusaurus#10556]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#issues).

### Motivation

# Docusaurus Faster

This is a meta-issue for the Docusaurus Faster project.

The goal is to modernize our infrastructure and greatly reduce the time it takes to build a Docusaurus site in production mode.

It includes multiple experimental PRs adding new `siteConfig.future.experimental_faster` options:

- [`swcJsLoader`](https://github.com/facebook/docusaurus/pull/10435)
- [`swcJsMinimizer`](https://github.com/facebook/docusaurus/pull/10441)
- [`swcHtmlMinimizer `](https://github.com/facebook/docusaurus/pull/10554)
- [`lightningCssMinimizer`](https://github.com/facebook/docusaurus/pull/10522)
- [`mdxCrossCompilerCache`](https://github.com/facebook/docusaurus/pull/10479)
- [`rspackBundler`](https://github.com/facebook/docusaurus/pull/10402)

We recommend turning them on all at once with `siteConfig.future.experimental_faster: true`, but you can also turn them on/off one by one to troubleshoot potential problems.

Note that the `rspackBundler` option requires `swcJsLoader`, `swcJsMinimizer`, and `lightningCssMinimizer` to be enabled first.

It is possible that certain combinations of options do not work, and our CI doesn't cover all possible permutations (only `true` / `false`).

## What to expect

Our Docusaurus website builds:
- ~5x faster on cold builds 🔥
- ~ same on hot re-builds (because [Rspack doesn't support persistent cache yet](https://rspack.dev/misc/planning/roadmap#persistent-cache))

Although rebuilds (with warm `./node_modules/.cache`) are not faster (yet), we recommend you enable the `experimental_faster: true` option to future-proof your site against our new dependencies and report us potential problems you encounter.

Our core plugins/themes are all compatible with this new option, and we expect most community plugins to be compatible out-of-the-box. Some plugins might need minor adjustments to make them compatible with Rspack, and we'll help the community to make our ecosystem compatible with Rspack.

This option is relatively safe to use in production: we have deployed our own website with it already:
- If your Docusaurus site is standard, it should be fine out of the box
- If your Docusaurus site is more complex, make sure to test it thoroughly and give us feedback

## Contributing to this issue

We want our community to share feedback on this Docusaurus Faster project. 

Please try to adopt it as part of the v3.6+ / canary releases, and tell us how it works for you. 

If you encounter problems, please let us know here, and we might create dedicated issues if needed.

This general issue is a great place to follow the Docusaurus Faster project overall, as we try to stabilize it and aim to make it the default for an upcoming major version.

### Self-service

- [ ] I'd be willing to do some initial work on this proposal myself.

**Comment by sserrata** [comment]
Thanks @slorber, sharing comparison results for our [OpenAPI plugin site](https://docusaurus-openapi.tryingpan.dev/). 

# Cold Build No Optimization

```
[PERF] Build > en > Bundling with webpack - 41.58 seconds!
[PERF] Build > en > SSG > Load App renderer > Load server bundle - 8.49 ms
[PERF] Build > en > SSG > Load App renderer > Server bundle size = 17.241 MB
[PERF] Build > en > SSG > Load App renderer > Evaluate server bundle - 225.13 ms
[PERF] Build > en > SSG > Load App renderer - 234.15 ms
[PERF] Build > en > SSG > Generate static files - 3.26 seconds!
[PERF] Build > en > SSG - 3.50 seconds!
[PERF] Build > en > postBuild() - 10.58 ms
[PERF] Build > en > Broken links checker - 28.01 ms
[SUCCESS] Generated static files in "build".
[PERF] Build > en - 46.38 seconds!
[PERF] Build - 46.38 seconds!
[INFO] Use `npm run serve` command to test your build locally.
✨  Done in 51.97s.
```

# Cold Build With Experimental Faster Enabled

```
[PERF] Build > en > Bundling with webpack - 23.44 seconds!
[PERF] Build > en > SSG > Load App renderer > Load server bundle - 7.16 ms
[PERF] Build > en > SSG > Load App renderer > Server bundle size = 17.767 MB
[PERF] Build > en > SSG > Load App renderer > Evaluate server bundle - 222.76 ms
[PERF] Build > en > SSG > Load App renderer - 230.31 ms
[PERF] Build > en > SSG > Generate static files - 1.82 seconds!
[PERF] Build > en > SSG - 2.06 seconds!
[PERF] Build > en > postBuild() - 9.49 ms
[PERF] Build > en > Broken links checker - 27.01 ms
[SUCCESS] Generated static files in "build".
[PERF] Build > en - 26.79 seconds!
[PERF] Build - 26.79 seconds!
[INFO] Use `npm run serve` command to test your build locally.
✨  Done in 32.52s.
```

**Comment by slorber** [comment]
Thanks for reporting these early results @sserrata 

Note that the Rspack PR hasn't been merged yet and if you use canary today you are still using Webpack. 

For plugin authors, it's possible that Rspack will not work immediately out of the box. Some plugins will have to adjust their code to add Rspack support, eventually requiring a minimum version of Docusaurus v3.6+ to do so.

Follow this issue to know when a guide for plugin authors is ready: https://github.com/facebook/docusaurus/issues/10572

**Comment by slorber** [comment]
The Rspack support has been merged: https://github.com/facebook/docusaurus/pull/10402

You can try canary `0.0.0-6101` or above to benefit from all the upcoming perf improvements.

Here's an example PR upgrading the React Native website: https://github.com/facebook/react-native-website/pull/4268


## How to benchmark


One way to benchmark this is to use [hyperfine](https://github.com/sharkdp/hyperfine) with an env variable:

```ts
const config = {
  future: {
    experimental_faster: (process.env.DOCUSAURUS_FASTER ?? 'true') === 'true';,
  },
}
```

For cold builds:

```bash
hyperfine --prepare 'yarn clear' --runs 5 'DOCUSAURUS_FASTER=false yarn build' 'DOCUSAURUS_FASTER=true yarn build'
```

For warm rebuilds:

```bash
DOCUSAURUS_FASTER=false yarn build
hyperfine --runs 5 'DOCUSAURUS_FASTER=false yarn build' 'DOCUSAURUS_FASTER=true yarn build'
```

**Comment by o-l-a-v** [comment]
Tested with a relatively large internal documentation site. Locally, Windows, Yarn v4.5.1, NodeJS v20.18.0

Used following command in PowerShell for timing the build:

```pwsh
yarn docusaurus clear; $Stopwatch = [System.Diagnostics.Stopwatch]::StartNew(); yarn build; $Stopwatch.Stop(); $Stopwatch.Elapsed.ToString()
```

* 3.5.2: 00:02:32.9615610
* Canary with `experimental_faster: true`: 00:01:10.1179920

**Comment by sedghi** [comment]
which version of docusaurus comes with this experimental config? since i can't get it to run in 3.5.2

```
[ERROR] Error: These field(s) ("future.experimental_faster",) are not recognized in docusaurus.config.js.
If you still want these fields to be in your configuration, put them in the "customFields" field.
See https://docusaurus.io/docs/api/docusaurus-config/#customfields
    at validateConfig (/Users/alireza/dev/admin/cornerstone3D.git.worktrees/beta/node_modules/@docusaurus/core/lib/server/configValidation.js:299:15)
    at loadSiteConfig (/Users/alireza/dev/admin/cornerstone3D.git.worktrees/beta/node_modules/@docusaurus/core/lib/server/config.js:39:62)
    at async Promise.all (index 1)
    at async loadContext (/Users/alireza/dev/admin/cornerstone3D.git.worktrees/beta/node_modules/@docusaurus/core/lib/server/site.js:37:97)
    at async loadSite (/Users/alireza/dev/admin/cornerstone3D.git.worktrees/beta/node_modules/@docusaurus/core/lib/server/site.js:136:21)
    at async createReloadableSite (/Users/alireza/dev/admin/cornerstone3D.git.worktrees/beta/node_modules/@docusaurus/core/lib/commands/start/utils.js:50:16)
    at async Command.start (/Users/alireza/dev/admin/cornerstone3D.git.worktrees/beta/node_modules/@docusaurus/core/lib/commands/start/start.js:22:28)
[INFO] Docusaurus version: 3.5.2
```

**Comment by societymartingale** [comment]
I ran some informal benchmarks on a Macbook M3 Pro building our Docusaurus site. I cleared the cache after each build. Seeing a 2.5x speedup.

```
existing build (s)
-----------------
70.2
72.2
72.0
71.5
72.5
average: 71.7s

experimental_faster build (s)
-----------------------------
31.7
27.4
27.0
27.6
27.8
average: 28.3s

speedup = 71.7 / 28.3 = 2.5
```

**Comment by lebalz** [comment]
Wanted to test it on a docusaurus page that uses stage 3 decorators. Is it (or will it be) possible to customize the configuration of swc? (I have no experience so far with swc, but what i understand from the [docs](https://swc.rs/docs/configuration/compilation), the following should work:

```json
{
  "jsc": {
    "parser": {
      "syntax": "typescript",
      "decorators": true
    }
  }
}
```
)

**Comment by slorber** [comment]
I'm not 100% sure but maybe adding a .swcrc would work?
https://swc.rs/docs/configuration/swcrc

**Comment by lebalz** [comment]
Thanks for the advice @slorber . I tried it, but somehow swc does not pick up the .swcrc configuration.

Here is a little repro:

https://codesandbox.io/p/devbox/sharp-bird-jd7tf5

With `future: { experimental_faster: { swcJsLoader: false } }` it compiles and works as intended (click the `Click Me` Button and check the console log to see it working).

The used code for the decorators can be compiled by the swc playground when enabling the decorators.

<details>
<summary>decorated code example</summary>

```ts
type AccessorDescriptor = {
  get: () => number;
  set: (value: number) => void;
};

type DecoratorContext = {
  name: string;
  kind: string;
};

function accessorDecorator(
  { get, set }: AccessorDescriptor,
  { name, kind }: DecoratorContext
) {
  if (kind === "accessor") {
    return {
      init(initialValue: any) {
        return initialValue ?? 0;
      },
      get() {
        const value = get.call(this);
        // Additional logic can be added here
        return value;
      },
      set(newValue: any) {
        const oldValue = get.call(this);
        console.log("set", name, "=", newValue);
        // Additional logic can be added here
        set.call(this, newValue);
      },
    };
  }
}

class ExampleClass {
  @accessorDecorator accessor exampleField: number = 1;
}

export default ExampleClass;
```
</details>

Maybe the options must be passed to swc explicitely 👉 https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-faster/src/index.ts#L25-L34 ?

**Comment by slorber** [comment]
I don't know, will need to investigate.

Note that you might still be able to use the old custom js loader option here: https://github.com/facebook/docusaurus/issues/4765#issuecomment-841135926

Didn't try, but pseudo-code:

```js
const config = {
  webpack: {
    jsLoader: (isServer) => ({
      loader: 'builtin:swc-loader', // (only works with Rspack)
      options: require("@docusaurus/faster").getSwcLoaderOptions({isServer}),
    }),
  },
}
```

**Comment by slorber** [comment]
Docusaurus v3.6 is out with the Docusaurus Faster options 🥳 🎉 

https://docusaurus.io/blog/releases/3.6

**Comment by o-l-a-v** [comment]
Would it be possible to add an option in `create-docusaurus` to add `@docusaurus/faster` and enable it by adding `future: {experimental_faster: true}` to the config?

* <https://docusaurus.io/docs/api/misc/create-docusaurus>

Maybe `--docusaurus-faster` with default value `false`?

**Comment by slorber** [comment]
@o-l-a-v I'm not sure it's necessary. Depending on initial feedback I plan to make it soon the default of our init template to increase adoption, feedback, bugfixes, and finally making it the new default infrastructure.

**Comment by ruibaby** [comment]
Hi, I found a new problem after turning on faster mode. This may be related to rspack. This problem only occurs in the development environment (pnpm start).

```bash
❯ pnpm start DOCUSAURUS_PERF_LOGGER=true

> @halo-dev/docs@0.0.0 start /Users/ryanwang/Workspace/github/ruibaby/docs
> docusaurus start "DOCUSAURUS_PERF_LOGGER=true"


  ----------------------------------------------------------------------------------------------------------------------------------------------- 
                                                                                                                                                  
                                                    Update available 3.6.0-canary-6132 → 3.6.0                                                    
                                                                                                                                                  
                                To upgrade Docusaurus packages with the latest version, run the following command:                                
       `npm i @docusaurus/core@latest @docusaurus/faster@latest @docusaurus/plugin-client-redirects@latest @docusaurus/preset-classic@latest      
     @docusaurus/theme-classic@latest @docusaurus/theme-common@latest @docusaurus/theme-mermaid@latest @docusaurus/module-type-aliases@latest`    
                                                                                                                                                  
  ----------------------------------------------------------------------------------------------------------------------------------------------- 

[INFO] Starting the development server...

[ERROR] [Error: ENOENT: no such file or directory, lstat '/Users/ryanwang/Workspace/github/ruibaby/docs/DOCUSAURUS_PERF_LOGGER=true'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'lstat',
  path: '/Users/ryanwang/Workspace/github/ruibaby/docs/DOCUSAURUS_PERF_LOGGER=true'
}
[INFO] Docusaurus version: 3.6.0-canary-6132
Node version: v22.10.0
 ELIFECYCLE  Command failed with exit code 1.
❯ DOCUSAURUS_PERF_LOGGER=true pnpm start

> @halo-dev/docs@0.0.0 start /Users/ryanwang/Workspace/github/ruibaby/docs
> docusaurus start


  ----------------------------------------------------------------------------------------------------------------------------------------------- 
                                                                                                                                                  
                                                    Update available 3.6.0-canary-6132 → 3.6.0                                                    
                                                                                                                                                  
                                To upgrade Docusaurus packages with the latest version, run the following command:                                
       `npm i @docusaurus/core@latest @docusaurus/faster@latest @docusaurus/plugin-client-redirects@latest @docusaurus/preset-classic@latest      
     @docusaurus/theme-classic@latest @docusaurus/theme-common@latest @docusaurus/theme-mermaid@latest @docusaurus/module-type-aliases@latest`    
                                                                                                                                                  
  ----------------------------------------------------------------------------------------------------------------------------------------------- 

[INFO] Starting the development server...
[PERF] Load site > Load context - 115.72 ms - (48mb -> 64mb)
[PERF] Load site > Load plugins > Init plugins - 191.18 ms - (64mb -> 78mb)
[PERF] Load site > Load plugins > Load plugins content > Load classic@default > translatePluginContent() - 8.34 ms - (79mb -> 81mb)
[PERF] Load site > Load plugins > Load plugins content > Load pages@default > loadContent() - 9.00 ms - (79mb -> 81mb)
[PERF] Load site > Load plugins > Load plugins content > Load pages@default - 19.53 ms - (79mb -> 80mb)
[PERF] Load site > Load plugins > Load plugins content > Load classic@default > getDefaultCodeTranslationMessages() - 11.98 ms - (81mb -> 81mb)
[PERF] Load site > Load plugins > Load plugins content > Load classic@default - 20.62 ms - (79mb -> 81mb)
[PERF] Load site > Load plugins > Load plugins content > Load docs@default > loadContent() - 138.61 ms - (79mb -> 86mb)
[PERF] Load site > Load plugins > Load plugins content > Load docs@default > contentLoaded() - 39.50 ms - (87mb -> 95mb)
[PERF] Load site > Load plugins > Load plugins content > Load docs@default - 180.91 ms - (79mb -> 95mb)
[PERF] Load site > Load plugins > Load plugins content - 181.09 ms - (79mb -> 95mb)
[PERF] Load site > Load plugins - 374.37 ms - (64mb -> 95mb)
[PERF] Load site > Create site files - 46.84 ms - (96mb -> 100mb)
[PERF] Load site - 538.29 ms - (48mb -> 100mb)
[SUCCESS] Docusaurus website is running at: http://localhost:3000/
● Client ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ (83%) sealing chunk ids                                                                Panic occurred at runtime. Please file an issue on GitHub with the backtrace below: https://github.com/web-infra-dev/rspack/issues
Message:  byte index 2 is not a char boundary; it is inside '务' (bytes 0..3) of `务端-1-d-0-f-04c`
Location: crates/rspack_core/src/utils/compile_boolean_matcher.rs:325

Backtrace omitted.

Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
 ELIFECYCLE  Command failed.
[1]    97491 abort      DOCUSAURUS_PERF_LOGGER=true pnpm start
```

**Comment by societymartingale** [comment]
I enabled `experimental_faster` in the CI pipeline, and it reduced the build job duration by 3.5 minutes. This is for a Docusaurus site with 1100 markdown documents. I really appreciate the work that went into these optimizations!

**Comment by mike-solomon** [comment]
To add to the list of wins - with faster enabled on one of my sites with 4651 markdown files, the build job went from taking up to 15 minutes to build down to only ~7.5 minutes 🥳 . Thanks so much for the effort y'all put into optimizing this.

**Comment by felipecrs** [comment]
The build from one internal website here went from 4m18s to 1m8s. A great deal, thank you very much!

**Comment by jdussouillez** [comment]
Migrated a small doc today very easily, no problem.

Build time went from 1m29s to 0m14s! :rocket: Thanks!

**Comment by ShangjinTang** [comment]
**Incredible and impressive improvement.**

Tested at same environment on same machine with clean build command `hyperfine "rm -rf build && yarn build" --runs=3`:

- Docusaurus 3.5.2:
    ```text
    Time (mean ± σ):     98.320 s ± 13.658 s
    Range (min … max):   87.976 s … 113.801 s
    ```

- Docusaurus 3.6.1 with config `future: { experimental_faster: true, }`:
    ```text
    Time (mean ± σ):     19.712 s ±  1.784 s
    Range (min … max):   18.664 s … 21.772 s
    ```

Took only 20% (19.712s / 98.320s) build time compared to previous version.

**Comment by karl-cardenas-coding** [comment]
Our Spectro Cloud documentation build time went from 1 hour and 15 minutes to 35 min 🎉

For context, our builds run on a self-hosted GitHub runner with 8 CPU 16 GB Memory and 150 GB SSD.

**Comment by jackw** [comment]
Our Grafana plugin-tools docs builds times went from 36s to 11s build time. Every little helps!

Really appreciate the efforts everyone made here. 👏👏👏

**Comment by slorber** [comment]
## Important security notice for Docusaurus Faster users

Version 1.1.7 of `@rspack/core` has been compromised with a malicious npm publish.

More info here:
https://github.com/web-infra-dev/rspack/releases/tag/v1.1.8

The 1.1.7 version was published yesterday. The Rspack team reacted quickly to deprecate it and publish a safe 1.1.8 version within 1h after the 1.1.7 malicious publish.

`@rspack/core` is installed as part of `@docusaurus/faster`, so if you installed `@docusaurus/faster` yesterday, or upgraded dependencies, make sure you are not using v1.1.7 of `@rspack/core`.

If you are unlucky and installed it yesterday during that 1h period, you may have installed that compromised version. This may also happen when upgrading dependencies or re-generating your package manager lock file.

You can check the `@rspack/core` version your site is using with package manager commands:
- `npm why @rspack/core`
- `yarn why @rspack/core`

Alternatively, you can look at `node_modules/@rspack/core/package.json` for the version field.


If your Docusaurus site is using Rspack v1.1.7, make sure to upgrade to v1.1.8 immediately, and check for unusual activity on your system.

More info here:
https://github.com/web-infra-dev/rspack/releases/tag/v1.1.8

---

Note that the large majority of Docusaurus users should not be affected:
- `@rspack/core` is not installed by default, it is only installed manually with `@docusaurus/faster`
- The Rspack team reacted within 1h so even if you use `@docusaurus/faster`, you really need to be unlucky if you have installed/resolved dependencies at the worst possible time yesterday

**Comment by LiuG-lynx** [comment]
Is this plugin forward compatible? Do older versions support this feature？

**Comment by CMCDragonkai** [comment]
This is pretty cool, would be great to see it soon.

I was hoping that there could be some optimisations to how assets are handled too, especially when referring to static assets that are also in the static folder. Right now seems like assets get duplicated between `static/...` and `public/assets/...`.

**Comment by alirezamirian** [comment]
Build time going from 3:20 build time to 0:24 🚀  
Amazing job, and a very nice approach to implement with the incremental API!
Is there any plan for allowing .swcrc the same way .babelrc was supported when using babel loader for js?
I needed to set `jsc.parser.decorators` to true and had to either patch `@docusaurus/faster` or disable `swcJsLoader` for now.

**Comment by slorber** [comment]
> Build time going from 3:20 build time to 0:24 🚀 Amazing job, and a very nice approach to implement with the incremental API! Is there any plan for allowing .swcrc the same way .babelrc was supported when using babel loader for js? I needed to set `jsc.parser.decorators` to true and had to either patch `@docusaurus/faster` or disable `swcJsLoader` for now.

thanks for reporting!

You should be able to customize the swc loader using our existing `siteConfig.webpack.jsLoader` config, as explained here: https://github.com/facebook/docusaurus/issues/10556#issuecomment-2454482074

Maybe using `.swcrc` already works. I haven't tried, but we didn't do anything to explicitly disable support of this file.

**Comment by slorber** [comment]
Docusaurus v3.8+ will introduce a new Docusaurus Faster option: [`ssgWorkerThreads `](https://github.com/facebook/docusaurus/pull/10826). 

Depending on your site, it could reduce the SSG time by 50%+ (even more for larger sites), and decrease total build time by more than 15%+.

If you have Docusaurus Faster enabled globally, it will be enabled by default when you upgrade to v3.8+.

---

This new option requires to turn on a Docusaurus v4 future flag: `future.v4. removeLegacyPostBuildHeadAttribute`: this permits to opt-in for a breaking change we plan in our plugin API in the upcoming major version of Docusaurus. In practice, this shouldn't cause any trouble, I haven't seen any community plugin using this API, and is relatively safe to turn on.

If you are using Docusaurus Faster, I recommend using the following settings:

```ts
export default {
  future: {
    v4: true, // Opt-in for upcoming v4 breaking changes
    experimental_faster: true, // Turn all Docusaurus Faster options on
  },
};
```

**Comment by karl-cardenas-coding** [comment]
@alirezamirian @slorber We ran into the same issue and disabled the flag `swcJsLoader: false` because for some odd reason, Netlify returns an exit code 2 when the flag is enabled. All other CI/CD systems, such as GitHub Actions, work fine. It's only in Netlify that we see this issue. 

I tried the workaround mentioned in the comment linked above and had no luck. Will this prevent us from upgrading to 3.8 in the future?

**Comment by slorber** [comment]
@karl-cardenas-coding 

> We ran into the same issue

The same issue? Which one exactly?

I never heard of such a timeout issue affecting only Netlify.

v3.8 doesn't enable Docusaurus Faster by default, it would be a breaking change that can only be done in an upcoming major version of Docusaurus. I'm not sure yet when we'll turn Docusaurus Faster on by default. During the transition, we'll eventually keep an option for users to downgrade to Webpack and "Docusaurus Slower".

**Comment by karl-cardenas-coding** [comment]
@slorber My bad, I wasn't specific. It's really odd, if we keep `swcJsLoader: true`, it fails only in Netlify (link to [logs](https://app.netlify.com/sites/docs-spectrocloud/deploys/67981157a1a96e00087aa42c). Locally and in our GitHub Action everything works fine. We don't get an error message in Netlify that is helpful.  

At first we thought it was because of the CSS deprecation warning, but that's not the case as we got those as well when `swcJsLoader: false` is set.  I've also tried different Netlify tricks, like setting the build command to `CI= ` to ignore build errors on warning and no luck. I'm really puzzled as to why our Netlify breaks when  `swcJsLoade` is set to `true` 

¯\_(ツ)_/¯

**Comment by taraspos** [comment]
Hi @slorber, 
I'm seeing following build when trying to upgrade from 3.6 to 3.7 (tried with `3.7.0-canary-6207` too):

```
Panic occurred at runtime. Please file an issue on GitHub with the backtrace below: https://github.com/web-infra-dev/rspack/issues
Message:  width 3 given for non-narrow character
Location: index.crates.io-6f17d22bba15001f/swc_common-5.0.0/src/syntax_pos.rs:714

Run with COLORBT_SHOW_HIDDEN=1 environment variable to disable frame filtering.
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
<empty backtrace>
Aborted
error Command failed with exit code 134.
```

- https://github.com/swc-project/swc/blob/37616c33bf877845afe55c9fc0d21ccbbf59bad3/crates/swc_common/src/syntax_pos.rs#L709-L716

I tried building with `NODE_ENV=production` to disable incremental builds and `SKIP_HTML_MINIFICATION=true` (as mentioned in https://github.com/facebook/docusaurus/issues/9985#issuecomment-2407755264) but it still fails.

You can reproduce it by trying to build:
- https://github.com/gravitational/docs-website/pull/122

**Comment by Zenahr** [comment]
@slorber I'm running into an issue with SVG files being minified so much so that even IDs disappear (both in dev and build).  
I've found a SO article walking through modifying the SVGO library config via webpack (the basis of SVGR which I think Docusaurus uses.

Is this possible with how SWC is currently set up in Docusaurus or would I need to switch back to Webpack to do this?

https://stackoverflow.com/a/78487098

**Comment by slorber** [comment]
@Zenahr this is not related to Docusaurus Faster so I'm going to hide these comments to keep the discussion on Docusaurus Faster features

To answer your questions:
- in v3.7 has a dedicated SVGR plugin letting you pass custom SVGO config: https://docusaurus.io/blog/releases/3.7#svgr-plugin
- for v4, we are considering improving our default SVGO config (breaking change): https://github.com/facebook/docusaurus/issues/10679

**Comment by slorber** [comment]
Docusaurus v3.8+ will introduce a new Docusaurus Faster option: [`rspackPersistentCache `](https://github.com/facebook/docusaurus/pull/10931). You can try it already when using the Docusaurus canary release channel.

It should help your site rebuild faster when using the Rspack bundler (requires `rspackBundler: true`)

To make it work, you also have to persist the `node_modules/.cache` folder across site build jobs (similar to webpack persistent cache support).

As of today, the perf improvement is not super impressive. On our website:
- warm site re-builds are ~25% faster
- the bundling phase on re-builds is ~33% faster

It's still an improvement though, but not comparable to the improvement we have on webpack where bundling improves by 300-800%.

We expect these results to improve significantly in the future once Rspack improves their persistent cache, and once they fix a persistent cache bug we reported. I'll post an update here if we see an improvement.

More benchmarks and details are available at https://github.com/facebook/docusaurus/pull/10931

**Comment by IntrepidToucan** [comment]
Hello, I've noticed that when we turn on the `experimental_faster.swcHtmlMinimizer` option, the built HTML doesn't have `<head></head>` tags or a closing `</body>` tag.

You can also see this in the Docusaurus website itself by viewing the static source (`view-source:https://docusaurus.io/docs`).

This malformed HTML caused an issue for us, so we fixed it by turning off `swcHtmlMinimizer`. Is this something that could be fixed in a future release?

**Comment by slorber** [comment]
> Docusaurus v3.8+ will introduce a new Docusaurus Faster option: [`rspackPersistentCache `](https://github.com/facebook/docusaurus/pull/10931). You can try it already when using the Docusaurus canary release channel.
> 
> It should help your site rebuild faster when using the Rspack bundler (requires `rspackBundler: true`)
> 
> To make it work, you also have to persist the `node_modules/.cache` folder across site build jobs (similar to webpack persistent cache support).
> 
> As of today, the perf improvement is not super impressive. On our website:
> 
> * warm site re-builds are ~25% faster
> * the bundling phase on re-builds is ~33% faster
> 
> It's still an improvement though, but not comparable to the improvement we have on webpack where bundling improves by 300-800%.
> 
> We expect these results to improve significantly in the future once Rspack improves their persistent cache, and once they fix a persistent cache bug we reported. I'll post an update here if we see an improvement.
> 
> More benchmarks and details are available at [#10931](https://github.com/facebook/docusaurus/pull/10931)

With the Rspack 1.3.6 upgrade, the persistent cache will be faster for Docusaurus sites. 

It will be included in the upcoming v3.8 release.

Updated benchmarks available here: https://github.com/facebook/docusaurus/pull/11114

TLDR: The impact of the persistent cache on total build times:
- Small site: 15.8s -> 9.3s (before the upgrade: 10.4s)
- Medium site: 37.4s -> 24.9s (before the upgrade: 29.7s)

The bundling phase is x2-x3 faster with the Rspack persistent cache. Bundling is still the most expensive phase of Docusaurus, but it's not the only expensive one to optimize.

**Comment by s-nix** [comment]
Edit: This was an OS-level issue. GH Actions build is running successfully.

Build times are down from approx. 5 minutes to 2. Great work!!! 😃
 

-----

Running into the following error throwing uncountable numbers of times when running `yarn build`

I've tried `yarn clean` and deleting `package-lock.json` and `yarn.lock` without success.

`Error: EMFILE: too many open files`

Relevant config details:

### `docusaurus.config.ts`
```ts
const config: Config = {
  future: {
    v4: true,
    experimental_faster: true,
  },
}
```

### `package.json`
```json
"dependencies": {
  "@docusaurus/core": "^3.7.0",
  "@docusaurus/faster": "^3.8.0",
}
```

### Number of Document/Static Files
```
Count Name
----- ----
 7758 .webp
 2546 .md
   43 .json
   33 .png
   19 .xml
   11 .svg
   11 .ps1
    6 .pdf
    4 .gif
```

**Comment by slorber** [comment]
@s-nix This doesn't look related to Docusaurus Faster unless proven otherwise. It's impossible to do anything with the provided information. It might be a generic Node error related to [your OS settings](https://stackoverflow.com/questions/8965606/node-and-error-emfile-too-many-open-files). 

Can you please open a dedicated issue/discussion, including a repro? (is this [proval-docs](https://github.com/ProVal-Tech/proval-docs)?) 

Edit: found it: https://github.com/ProVal-Tech/proval-docs/issues/178

**Comment by s-nix** [comment]
@slorber You were correct that this was OS-related. Either it's just Windows in general or some other intrusive security application that may or may not have been recently forced upon us. Either way, excellent work on Faster. Build times are down over 50%.

**Comment by ramnivas** [comment]
> Hello, I've noticed that when we turn on the `experimental_faster.swcHtmlMinimizer` option, the built HTML doesn't have `<head></head>` tags or a closing `</body>` tag.
> 
> You can also see this in the Docusaurus website itself by viewing the static source (`view-source:https://docusaurus.io/docs`).
> 
> This malformed HTML caused an issue for us, so we fixed it by turning off `swcHtmlMinimizer`. Is this something that could be fixed in a future release?
> 
> **Edit from [@slorber](https://github.com/slorber):** answered here [#10554 (comment)](https://github.com/facebook/docusaurus/pull/10554#issuecomment-2775197012)

One effect of not having `<head>`: LinkedIn's OG image (testable at https://www.linkedin.com/post-inspector/inspect) doesn't work. I had `swcHtmlMinimizer` set to true (due to setting of `experimental_faster` to true). It took quite a bit of time to figure the real issue (the internet is full of advice on how to fix LinkedIn's OG image problems, so that led to a wild chase into nothing!). Maybe this can be fixed or documented?


---

Edit @slorber: this has been fixed in https://github.com/facebook/docusaurus/pull/11383

**Comment by slorber** [comment]
Thanks @ramnivas, I didn't know that this would impact the LinkedIn OG image. That would be great if LinkedIn would support that, but I guess it would be hard to ask them, and other platforms could behave the same anyway.

I don't think SWC html minimizer permits to disable the removal of `<head>`, so maybe we should open an issue there to ask for that feature, cc @kdy1

**Comment by kdy1** [comment]
I can add an option. Can you file an issue on the main swc repository?

**Comment by karl-cardenas-coding** [comment]
Hey @slorber - I may have stumbled on a bug in the `loadMDXWithCaching` function  in the context of hot-module reload. I am consistently getting the error message 

```shell
[ERROR] Error: this should never be called
    at noop (/Users/workUser/projects/example/abc/docs/node_modules/@docusaurus/mdx-loader/lib/loader.js:125:23)
```

If I disable `future.experimental_faster.mdxCrossCompilerCache` then hot-reloading works fine. I took a peek at the function and I think the issue stems from the assumption that server will always compile first. 

```jsx
if (compilerName === 'client') {
    const promise = loadMDX({
        fileContent,
        filePath,
        options,
        compilerName,
    });
    if (cacheEntry) {
        // Server was already waiting - resolve its promise
        promise.then(cacheEntry.resolve, cacheEntry.reject);
        deleteCacheEntry();
    }
    else {
        // ❌ BUG: Assumes client NEVER compiles first
        const noop = () => {
            throw new Error('this should never be called');
        };
        crossCompilerCache.set(cacheKey, {
            promise,
            resolve: noop, 
            reject: noop,   
        });
    }
    return promise;
}
```
I played with a local fix to see if I could get things to work.
The fix below changes the behavior when the client compiles first. Instead of throwing errors with placeholder "noop" functions, I added a deferred promise that the server can wait on. When the client finishes compiling, the result is forwarded to that promise. 

```jsx
     if (compilerName === 'client') {
         const promise = loadMDX({
             fileContent,
             filePath,
             options,
             compilerName,
         });
         if (cacheEntry) {
             promise.then(cacheEntry.resolve, cacheEntry.reject);
             deleteCacheEntry();
         }
         else {
-            const noop = () => {
-                throw new Error('this should never be called');
-            };
-            crossCompilerCache.set(cacheKey, {
-                promise,
-                resolve: noop,
-                reject: noop,
-            });
+            // Client first (during HMR) - create deferred promise for server
+            const deferred = (0, utils_2.promiseWithResolvers)();
+            crossCompilerCache.set(cacheKey, deferred);
+            promise.then(deferred.resolve, deferred.reject);
         }
         return promise;
     }
```

But I don't have the entire context if what I have is right but figured I run it by you 😅

**Comment by slorber** [comment]
@karl-cardenas-coding can you please post all those details in the PR dedicated to the cross compiler cache here? https://github.com/facebook/docusaurus/pull/10479

I believe this is an integration problem. The cross-compiler cache shouldn't be used in dev, for hot reloading, since there's no server in dev. Unfortunately, you don't share enough about your site, the plugins used, and which command you run, for me to fully understand the problem.

**Comment by OfficialMutazAwad** [comment]
New reply to your ticket #13102

Title: Re: [facebook/docusaurus] Docusaurus Faster (Issue #10556)

Message: Hello,we received your message and our support will reach out to you as soon as possible.

**Comment by karl-cardenas-coding** [comment]
@slorber will do. I'll see if I can sneak in some time today or tomorrow to reproduce this for you.

**Comment by slorber** [comment]
Your site uses one of those?
- the `showLastUpdateAuthor` and `showLastUpdateTime` plugin options
- the sitemap `lastmod` option

We have found these options to be a significant bottleneck for many sites due to issuing thousands of `git log <filename>` calls.

---

🆕 In Docusaurus v3.10, we will release:
- A new VCS API (Version Control System) to give you flexibility to read file update info on your own terms.
- A new Git-based "git-eager" implementation that reads the whole repository eagerly on `docusaurus build` instead of issuing thousands of `git log <filename>` calls.


More details on the PR: https://github.com/facebook/docusaurus/pull/11512

The impact on build times can be significant, in particular for:

- large Git repositories
- large number of md/mdx files
- large number of i18n locales
- all 3 above combined!

**Comment by OfficialMutazAwad** [comment]
New reply to your ticket #13123

Title: Re: [facebook/docusaurus] Docusaurus Faster (Issue #10556)

Message: Hello,we received your message and our support will reach out to you as soon as possible.

**Comment by karl-cardenas-coding** [comment]
@slorber, I tried all day to reproduce the issue in a new project, and I simply cannot. To my surprise, I am also unable to reproduce in our original code base. To answer your question above, none of those settings are actually used. I'll keep an eye on it. For now, I'm gonna re-enable `mdxCrossCompilerCache` and see if it happens again.  One thing I realized in this excercise is that my rspack was on `Rspack 1.5.7.` despite being on the latest Docusaurus version.

**Comment by tats-u** [comment]
If `docusaurus start` or `docusaurus build` is not finished and is stuck at "100% emitting after emit" in Windows with Rspack 1.6.6–1.7.3, try updating Rspack (`@rspack/core`) to 1.7.4 or later. (I could not find out why this happens) This is suppressed if you disable the persistent cache.

**Comment by AlbertBrand** [comment]
> [@alirezamirian](https://github.com/alirezamirian) [@slorber](https://github.com/slorber) We ran into the same issue and disabled the flag `swcJsLoader: false` because for some odd reason, Netlify returns an exit code 2 when the flag is enabled. All other CI/CD systems, such as GitHub Actions, work fine. It's only in Netlify that we see this issue.

I'm seeing a build problem, only when running on GitLab and when `swcJsLoader` is set to `true`:

```
        Error: Can't render static file for pathname "/[REDACTED]"
            at generateStaticFile ([REDACTED]/node_modules/.pnpm/@docusaurus+core@3.9.2_@docusaurus+faster@3.9.2_@docusaurus+types@3.9.2_@swc+core@1.15._9ec5c227eab01026f6cdca42e873a4f8/node_modules/@docusaurus/core/lib/ssg/ssgRenderer.js:131:20)
            at async [REDACTED]/node_modules/.pnpm/p-map@4.0.0/node_modules/p-map/index.js:57:22 {
          [cause]: ReferenceError: Prism is not defined
              at 88020 (server.bundle.js:2464:3)
```

Especially the `Prism is not defined` feels like an import/transpilation issue inside `server.bundle.js`. And the stacktrace is not relevant. Strange thing that I can't reproduce it in a local Docker container, it might be a CI env setting that influences the swc build process?

**Comment by slorber** [comment]
FYI, Docusaurus Faster is now stable (https://github.com/facebook/docusaurus/pull/11802).

It will be enabled by default in Docusaurus v4, and is also used for all newly initialized Docusaurus v3 sites.

We now have a v4 future flag so that users can prepare their site for Docusaurus v4: `future.v4: true` or `future.v4.fasterByDefault: true`.

In Docusaurus v4, we'll likely introduce a "Docusaurus Slower" mode, requiring a `@docusaurus/slower` dependency, for those who still want to use webpack/Babel. 

We'll likely drop support for the legacy infrastructure in Docusaurus v5, so please let us know if you are unable to upgrade your site to Docusaurus Faster for any reason! 
(Please open a bug report with a minimal repro, not a full/complex OSS repo)

**Comment by AlbertBrand** [comment]
> I'm seeing a build problem, only when running on GitLab and when `swcJsLoader` is set to `true`:
> 
> ```
>         Error: Can't render static file for pathname "/[REDACTED]"
>             at generateStaticFile ([REDACTED]/node_modules/.pnpm/@docusaurus+core@3.9.2_@docusaurus+faster@3.9.2_@docusaurus+types@3.9.2_@swc+core@1.15._9ec5c227eab01026f6cdca42e873a4f8/node_modules/@docusaurus/core/lib/ssg/ssgRenderer.js:131:20)
>             at async [REDACTED]/node_modules/.pnpm/p-map@4.0.0/node_modules/p-map/index.js:57:22 {
>           [cause]: ReferenceError: Prism is not defined
>               at 88020 (server.bundle.js:2464:3)
> ```

Figured this out. 1) Run `docusaurus clear` locally to make sure environments are identical and 2) this is triggered by [a bug in Redocusaurus.](https://github.com/rohit-gohri/redocusaurus/issues/388). Fix is incoming.
